src/main/java/javaslang/collection/Vector.java

```diff
@@ -765,12 +773,20 @@ public int length() {
 
     @Override
     public Vector<T> prepend(T element) {
-        return insert(0, element);
+        final int newIndexShift = indexShift - 1;
+        return new Vector<>(newIndexShift, trie.put(newIndexShift, element));
     }
 
     @Override
     public Vector<T> prependAll(java.lang.Iterable<? extends T> elements) {
-        return insertAll(0, elements);
+        List<T> list = List.ofAll(elements);
```

Comment:
mmhh, we could convert the elements to an IndexedSeq. If it is already an indexed seq (checked via instanceof) it is just a cast in O(1), otherwise we could put the elements in an Object[] and wrap it in an Array.

Later in the for-loop we do a indexedSeq.get(i) instead of list.head().

Don't know if that is really better...